proc_macros/src/lib.rs

```diff
@@ -165,3 +203,599 @@ pub fn rule_tests(input: TokenStream) -> TokenStream {
+            initializer,
+        })
+    }
+}
+
+struct RuleStateScopeSection {
+    scope: RuleStateScope,
+    fields: Vec<RuleStateFieldSpec>,
+}
+
+impl Parse for RuleStateScopeSection {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let scope: RuleStateScope = input.parse()?;
+        let mut fields: Vec<RuleStateFieldSpec> = Default::default();
+        while !input.is_empty() && !input.peek(token::Bracket) {
+            fields.push(input.parse()?);
+            if !input.is_empty() {
+                let _ = input.parse::<Token![,]>();
+            }
+        }
+        Ok(Self { scope, fields })
+    }
+}
+
+struct RuleStateSpec {
+    scope_sections: Vec<RuleStateScopeSection>,
+}
+
+impl Parse for RuleStateSpec {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let mut scope_sections: Vec<RuleStateScopeSection> = Default::default();
+        let rule_state_spec_content;
+        braced!(rule_state_spec_content in input);
+        while !rule_state_spec_content.is_empty() {
+            scope_sections.push(rule_state_spec_content.parse()?);
+            if !rule_state_spec_content.is_empty() {
+                let _ = rule_state_spec_content.parse::<Token![,]>();
+            }
+        }
+        Ok(Self { scope_sections })
+    }
+}
+
+struct RuleListenerSpec {
+    query: Expr,
+    capture_name: Option<Expr>,
+    callback: ExprClosure,
+}
+
+impl Parse for RuleListenerSpec {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let query: Expr = input.parse()?;
+        input.parse::<Token![=>]>()?;
+        let mut callback: Option<ExprClosure> = Default::default();
+        let mut capture_name: Option<Expr> = Default::default();
+        match input.parse::<ExprClosure>() {
+            Ok(parsed) => callback = Some(parsed),
+            _ => {
+                let content;
+                braced!(content in input);
+                while !content.is_empty() {
+                    let key: Ident = content.parse()?;
+                    content.parse::<Token![=>]>()?;
+                    match &*key.to_string() {
+                        "callback" => {
+                            callback = Some(content.parse()?);
+                        }
+                        "capture_name" => {
+                            capture_name = Some(content.parse()?);
+                        }
+                        key => panic!("Unexpected key: '{}'", key),
+                    }
+                    if !content.is_empty() {
+                        content.parse::<Token![,]>()?;
+                    }
+                }
+            }
+        }
+        Ok(Self {
+            query,
+            callback: callback.expect("Expected 'callback'"),
+            capture_name,
+        })
+    }
+}
+
+struct Rule {
+    name: Expr,
+    fixable: Option<Expr>,
+    state: Option<RuleStateSpec>,
+    listeners: Vec<RuleListenerSpec>,
+}
+
+impl Rule {
+    pub fn name_string(&self) -> String {
```

Comment:
This is a little wonky, the things it's returning are in two different casings and that's "working" because all of the consumers of this are then converting it to a certain case explicitly